File: src/main/java/com/hzcard/syndata/config/autoconfig/CanalClientContext.java
package com.hzcard.syndata.config.autoconfig;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

import com.hzcard.syndata.extractlog.Slaver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.context.ApplicationContext;


;

public class CanalClientContext implements Serializable, DisposableBean {

    private static Logger logger = LoggerFactory.getLogger(CanalClientContext.class);


    /**
     *
     */
    private static final long serialVersionUID = 1L;


    private CanalClientContext() {
    }

    private Map<String, Slaver> clients = new HashMap<>();

    /**
     * reset endPoint 与 repository 的class映射
     */
    private Map<String, Class> resetEndPoint = new ConcurrentHashMap<String, Class>();

    /**
     * repository 的class与 destination映射
     */
    private Map<Class, String> repositoryDestination = new ConcurrentHashMap<Class, String>();

    private ApplicationContext applicationContext;

    private CanalClientProperties canalClientProperties;


    public CanalClientProperties getCanalClientProperties() {
        return canalClientProperties;
    }


    public CanalClientContext(ApplicationContext applicationContext, CanalClientProperties canalClientProperties) {
        this.canalClientProperties = canalClientProperties;
        this.applicationContext = applicationContext;
    }

    public ApplicationContext getApplicationContext() {
        return this.applicationContext;
    }

//	public AbstractCanalClientScala getCanalClient(String destination){
//		return clients.get(destination);
////	}

    public boolean isExsits(String destination) {
        return clients.containsKey(destination);
    }

//	protected AbstractCanalClientScala regiestSimpleCanalClient(String destination,AbstractCanalClientScala client){
//		logger.error("client length is:{}",this.clients.size());
//		return this.clients.put(destination,client);
//	}

    public Slaver regiestSlaver(String destination, Slaver slaver) {
        return this.clients.put(destination, slaver);
    }

    @Override
    public void destroy() throws Exception {
        logger.error("canalclientContext destroy");
        for (Slaver client : clients.values()) {
            client.disconnect();
        }
    }

    /**
     * 重启slaver端
     * @param myChannel
     * @throws Exception
     */
    public void restart(String myChannel) throws Exception {
        clients.get(myChannel).disconnect();
        TimeUnit.SECONDS.sleep(5L);
        while(!clients.get(myChannel).connect()){}
    }

    public void start() {
        for (Slaver client : clients.values()) {
            client.connect();
        }
    }

    protected Class regiestResetEndPoint(String resetEndPoint, Class repository) {
        return this.resetEndPoint.computeIfAbsent(resetEndPoint, k -> repository);
    }

    public Class getRepositoryByEndPoint(String resetEndPoint) {
        return this.resetEndPoint.get(resetEndPoint);
    }

    public String getDestinationByRepository(Class repository) {
        return this.repositoryDestination.get(repository);
    }

//	protected void startClient() {
//		for(AbstractCanalClientScala client :this.clients.values()){
//			logger.error("canal client start");
//			client.start();
//		}
//
//	}

}

File: src/main/scala/com/hzcard/syndata/CanalClientContext.scala
package com.hzcard.syndata

import java.util
import java.util.Map
import java.util.concurrent.ConcurrentHashMap
import javax.annotation.{PostConstruct, PreDestroy}

import akka.actor.{ActorRef, ActorSystem}
import com.hzcard.syndata.config.autoconfig.CanalClientProperties
import com.hzcard.syndata.extractlog.Slaver
import com.hzcard.syndata.redis.RedisCache
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.{Autowired, Qualifier}
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.{DependsOn, Scope}
import org.springframework.stereotype.Component
import org.springframework.util.{Assert, ClassUtils}

/**
  * Created by zhangwei on 2017/7/18.
  */
@Component
@Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
@DependsOn(Array("actorSystem", "redisCache","mapDataSource","clusterActorRef"))
class CanalClientContext(@Autowired redis: RedisCache, @Autowired actorSystem: ActorSystem,
                         @Autowired canalClientProperties: CanalClientProperties,
                         @Autowired applicationContext: ApplicationContext) {
  private val logger = LoggerFactory.getLogger(classOf[CanalClientContext])

  import scala.collection.JavaConversions._
  import scala.collection.JavaConverters._

  private val resetEndPoint: Map[String, Class[_]] = new ConcurrentHashMap[String, Class[_]]
  private val clients: Map[String, Slaver] = new util.HashMap[String, Slaver]

  validate(canalClientProperties)
  canalClientProperties.getSchemas.asScala.filter(x => x._2 != null).
    filter(x => x._2.getTableRepository != null).
    map(x => x._2.getTableRepository).
    flatMap(x => x.values).
    filter(x => x.getRepository != null).
    foreach(x => {
      resetEndPoint.computeIfAbsent(x.getResetEndPoint, new java.util.function.Function[String, Class[_]] {
        override def apply(t: String): Class[_] = ClassUtils.forName(x.getRepository, applicationContext.getClassLoader)
      })
    })
  canalClientProperties.getDestinations.asScala.
    foreach(x => clients.put(x._1, new Slaver(x._2, actorSystem, applicationContext, redis)))
  start()

  @PreDestroy
  def cleanUp(): Unit = {
    System.out.println("canalclientContext destroy")
    import scala.collection.JavaConversions._
    for (client <- clients.values) {
      client.disconnect
    }
  }

  def validate(canalClientProperties: CanalClientProperties) = {
    Assert.notNull(canalClientProperties.getDestinations, "Destinations must not null")
    Assert.notEmpty(canalClientProperties.getSchemas, "table Repository must not null")
  }

  def getRepositoryByEndPoint(reporsitory: String) = {
    resetEndPoint.get(reporsitory)
  }

  def start(): Unit = {
    clients.asScala.foreach(x => x._2.connect())
  }

}
